src/Test/NodeTestCase.php

```diff
@@ -51,24 +68,58 @@ protected function getCompiler(?Environment $environment = null)
         return new Compiler($environment ?? $this->getEnvironment());
     }
 
+    /**
+     * @final since Twig 3.13
+     */
     protected function getEnvironment()
     {
-        if (!$this->currentEnv) {
-            $this->currentEnv = new Environment(new ArrayLoader());
-        }
+        return $this->currentEnv ??= static::createEnvironment();
+    }
 
-        return $this->currentEnv;
+    protected static function createEnvironment(): Environment
+    {
+        return new Environment(new ArrayLoader());
     }
 
+    /**
+     * @deprecated since Twig 3.13, use createVariableGetter() instead.
+     */
     protected function getVariableGetter($name, $line = false)
+    {
+        trigger_deprecation('twig/twig', '3.13', 'Method "%s()" is deprecated, use "createVariableGetter()" instead.', __METHOD__);
+
+        return self::createVariableGetter($name, $line);
+    }
+
+    final protected static function createVariableGetter(string $name, bool $line = false): string
     {
         $line = $line > 0 ? "// line $line\n" : '';
 
         return \sprintf('%s($context["%s"] ?? null)', $line, $name);
     }
 
+    /**
+     * @deprecated since Twig 3.13, use createAttributeGetter() instead.
+     */
     protected function getAttributeGetter()
+    {
+        trigger_deprecation('twig/twig', '3.13', 'Method "%s()" is deprecated, use "createAttributeGetter()" instead.', __METHOD__);
+
+        return self::createAttributeGetter();
+    }
+
+    final protected static function createAttributeGetter(): string
     {
         return 'CoreExtension::getAttribute($this->env, $this->source, ';
     }
+
+    /** @beforeClass */
+    #[BeforeClass]
+    final public static function checkDataProvider(): void
+    {
+        $r = new \ReflectionMethod(static::class, 'getTests');
+        if ($r->getDeclaringClass()->getName() !== self::class) {
+            trigger_deprecation('twig/twig', '3.13', 'Overriding "%s::getTests()" in "%s" is deprecated, override "provideTests()" instead.', self::class, static::class);
```

Comment:
As those methods are abstract, should we say "implementing" instead?

```suggestion
            trigger_deprecation('twig/twig', '3.13', 'Implementing "%s::getTests()" in "%s" is deprecated, implement "provideTests()" instead.', self::class, static::class);
```

Reply:
Good idea. I've adjusted the wording.